### DISCUSSION TITLE
Unable to access Embedded CFF Fonts

### DISCUSSION THREAD

**Discussion by alexpolmans** [airsdk/Adobe-Runtime-Support#826] [Q&A]
I have just upgraded AIR from 26 to 33, and cannot get embedded fonts to work anymore.

This code has worked for years

 ```actionscript
           var fonts : Array = Font.enumerateFonts(false);
            for each ( var item : Font in fonts )
            {
                if ( item.fontName == "My Embedded" && item.fontType == "embeddedCFF" )
                    return item;
            } // end for loop
```
but now the font type embeddedCFF seems to have changed.

Does anyone know how to find them?

**Comment by ajwfrost** [comment]
It's working for me with 33.1...
```
		[Embed(source = "ITCBLKAD.ttf", fontName = "BlackAdder Test", mimeType = "application/x-font", embedAsCFF = "true")] private var BlackAdderFont : Class;
		
		public function Main() 
		{
			var fontList : Array = Font.enumerateFonts();
			for (var i : uint = 0; i < fontList.length; i++)
			{
				trace("Font " + i + " = " + fontList[i].fontName + ", " + fontList[i].fontType);
			}
		}
```
Output:
```
Font 0 = BlackAdder Test, embeddedCFF
```
Are you able to provide the full test case code/swf?

**Reply by alexpolmans** [reply]
Thank you Andrew.  I cant send a tff file in this forum so I'll email you directly.

Another issue when I recreated my sdk was that the {airHome} variable was not defined.  I set it to ".." in the env.properties file.  Could it possibly be pointing ti the wrong place?

**Reply by ajwfrost** [reply]
Okay sure - if you zip it, it may attach.. or email me at andrew.frost at harman.com

For the {airHome} variable, this is a Flex SDK thing where I believe the installer should set the value but the raw SDK download then leaves it open. So you may just need to define it manually; I believe the path-element values are relative to the frameworks folder. If you're combining SDKs then yes the ".." may work but I'm not 100% sure how that part works - that's for the Flex compiler to pick up the SWC files. If it's wrong though, you'd know about it as things wouldn't compile at all!

thanks

**Reply by ajwfrost** [reply]
Hi @alexpolmans 

I've just seen your test case and yes we get the same as you: the `fontType` property is just displaying as `embedded` even when you have the `embedAsCFF:true;` style.

I've also been checking back to earlier versions of the AIR SDK but am still seeing the same thing. So, can you confirm that you did have it working with `embeddedCFF` showing up as the `fontType` with an earlier SDK?

This may actually be a Flex issue rather than AIR: there are some reports that the font manager wasn't working properly in recent Flex builds: e.g. https://stackoverflow.com/questions/23447407/spark-embedding-of-fonts-for-label

So, in your working SDK, what's the value in `flex-sdk-description.xml`? It might be possible to copy some of the tools across, or better still to keep the same Flex SDK and overlay this with the updated AIR SDK (after taking a back-up, of course..)

thanks

**Comment by alexpolmans** [comment]
Thanks Andrew,

I had to rebuild my dev environment completely and I know it’s not the same as it was before.  This was working just a week ago before everything crashed.  I probably do not have the save Flex SDK – with it being stored in the Program Files directly it was not backed up (I thought my whole PC was being backed up, and the IT department was only backing up the documents and desktop). When I got this computer it was easy to just download the Flex and AIR SDKs from the Adobe website, but now I see no trace of them.  So the only SDK I can find is the one from Apache (which I use in a different project). My last good build was on Flex SDK 16.0 and AIR SDK 29 from your website.

My current Flex SDK (from the install of Flash Builder) in flex-sdk-description.xml is 4.6.0.

We have three large projects dependent on AIR, in all about 600K lines of code with thousands of copies in the field, so I have to get it working. And this problem is only half of it.  A lot of the deployed units are black boxes running Linux, to I have to find Flex SDK 4.5.1 and AIR SDK 2.6 as well. Which is why I’m very interested in you getting Linux back on track.

So, if you have a copy of these SDKs, or know where I can find them, I will be seriously grateful.
               AIR SDK 2.6
               Flex SDK 4.5.1
               Flex SDK 16.0

Regards

Alex

**Comment by KarlSchumann** [comment]
A little late, but I believe I came across the same issue you are experiencing and was able to resolve it by using some files from a previous SDK and bringing that into my new SDK environment.
 - Apache FLEX 4.16.1
 - AIR 33.1.1.575

[This page](http://flex.apache.org/download-source.html) for Apache FLEX references the files needed in the optional dependencies related to font embedding. If I remember correctly using the FLEX downloader used to have a checkbox to allow you to download the files necessary for font embedding when it downloaded. Those checkboxes no longer appear as it sounds like it falls under a different license and so the optional dependencies don't download. That page has a link to an old 4.6.0.23201B version of the FLEX SDK which has the files in the {sdk}/lib folder

I had a previous SDK folder that must have been created using the download and that had the files here {sdk}/lib/external/optional so I kept them there instead of in the main lib folder, but seems like both should work.
 - afe.jar
 - aglj40.jar
 - flex-fontkit.jar
 - rideau.jar
 - swfutils.jar
